### PR TITLE
No uses for Executors

### PR DESCRIPTION
**Pull request by addisoncrump**
This formalizes some of the hooks concepts s.t. we can "prove" the safety of some of our global data operations.

### CODE REVIEW

File: libafl/src/executors/inprocess/mod.rs [hunk 1]
```diff
@@ -41,150 +41,162 @@ pub mod inner;
...
 }
 
-impl<EM, H, HB, HT, OT, S, Z> Executor<EM, Z> for GenericInProcessExecutor<H, HB, HT, OT, S>
+/// This function marks the boundary between the fuzzer and the target for in-process executors
+///
+/// # Safety
+/// This function sets a bunch of raw pointers in global variables, reused in other parts of
+/// the code.
+#[inline]
+pub fn enter_target<EM, I, S, Z>(
```
Comment (domenukk) [inline libafl/src/executors/inprocess/mod.rs:118]: This should be unsafe

File: libafl/src/executors/inprocess/mod.rs [hunk 2]
```diff
@@ -41,150 +41,162 @@ pub mod inner;
...
+    }
+}
+
+/// This function marks the boundary between the fuzzer and the target
+#[inline]
+pub fn leave_target() {
+    unsafe {
+        let data = addr_of_mut!(GLOBAL_STATE);
+
+        write_volatile(addr_of_mut!((*data).current_input_ptr), ptr::null());
```
Comment (domenukk) [inline libafl/src/executors/inprocess/mod.rs:156]: Clearly also unsafe, right?

File: libafl/src/executors/inprocess/mod.rs (hunk 1 above)
Comment (addisoncrump) [reply libafl/src/executors/inprocess/mod.rs:118]: Yup, this is going to disappear soon anyways (reimplementing this entirely in hooks/callbacks)

File: libafl/src/executors/inprocess/mod.rs (hunk 2 above)
Comment (addisoncrump) [reply libafl/src/executors/inprocess/mod.rs:156]: Yup, but we'll be changing this out soon anyways.

File: libafl/src/executors/hooks/inprocess.rs [hunk 3]
```diff
@@ -1,484 +1,839 @@
...
+}
 
-    /// Call before running a target.
-    #[allow(clippy::unused_self)]
-    #[allow(unused_variables)]
-    fn pre_exec(&mut self, state: &mut S, input: &S::Input) {
-        #[cfg(feature = "std")]
+impl<'a, E, EM, I, S, Z> GlobalContextGuard<'a, E, EM, I, S, Z> {
+    /// Access the members of the global access.
+    pub fn access(&mut self) -> (&mut E, &mut Z, &mut S, &mut EM, &I) {
```
Comment (domenukk) [inline libafl/src/executors/hooks/inprocess.rs:269]: leave a `Safety` comment on why this is safe

Comment (domenukk) [reply libafl/src/executors/hooks/inprocess.rs:269]: Also, the doc comment is confused

Comment (domenukk) [reply libafl/src/executors/hooks/inprocess.rs:269]: also, should it be called `get_mut()` or something?

**Comment by tokatoka** [comment]
ready to merge?

**Comment by addisoncrump** [comment]
Not nearly, no. I still need to update big chunks of the executor hooks.

**Pull request closed without merging**